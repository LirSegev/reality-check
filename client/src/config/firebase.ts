const firebaseConfig = {
	apiKey: 'AIzaSyBcPUPtr_37iqBlCUHpJhZy5cP8FBKqNY4',
	authDomain: 'reality-check-d6fe7.firebaseapp.com',
	databaseURL: 'https://reality-check-d6fe7.firebaseio.com',
	projectId: 'reality-check-d6fe7',
	storageBucket: 'reality-check-d6fe7.appspot.com',
	messagingSenderId: '243239923760',
	appId: '1:243239923760:web:f6bc0b9da0893fa7',
};
export default firebaseConfig;
