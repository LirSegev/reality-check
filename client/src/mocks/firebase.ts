import { mockFirebase } from 'ts-mock-firebase';

const firebase = mockFirebase();

export default firebase;

export const app = firebase.initializeApp({});
