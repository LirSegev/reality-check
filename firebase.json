{
	"functions": {
		"predeploy": [
			"npm --prefix \"$RESOURCE_DIR\" run lint",
			"npm --prefix \"$RESOURCE_DIR\" run build"
		],
		"source": "functions"
	},
	"hosting": {
		"site": "realitycheck",
		"public": "client/build",
		"ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
		"rewrites": [
			{
				"source": "**",
				"destination": "/index.html"
			}
		],
		"predeploy": ["npm --prefix ./client run build"]
	},
	"firestore": {
		"rules": "firestore.rules",
		"indexes": "firestore.indexes.json"
	}
}
