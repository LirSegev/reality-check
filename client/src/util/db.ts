import * as firebase from 'firebase/app';
import { store } from '../index';
import { Player } from './db.types';

/**
 * Updates the data of the current player in the database.
 * @param data An object containing the fields and values with which to update the document. Fields can contain dots to reference nested fields within the document.
 * @returns A Promise resolved once the data has been successfully written to the backend (Note that it won't resolve while you're offline).
 */
export async function updateCurrentPlayer(
	data: firebase.firestore.UpdateData
): Promise<void> {
	try {
		(await getCurrentPlayerRef()).update(data);
	} catch (err) {
		console.error(err);
		throw new Error('Error updating current user');
	}
}

/**
 * Gets the data of the current player from the database.
 */
export async function getCurrentPlayer(): Promise<Player | undefined> {
	try {
		const ref = await getCurrentPlayerRef();
		const doc = await ref.get();
		return doc.data() as Player | undefined;
	} catch (err) {
		console.error(err);
		throw new Error('Error getting current user');
	}
}

export function getCurrentPlayerRef(
	_getGameDocRef: typeof getGameDocRef = getGameDocRef
): Promise<firebase.firestore.DocumentReference> {
	return new Promise((resolve, reject) => {
		_getGameDocRef()
			.collection('players')
			.where('uid', '==', firebase.auth().currentUser!.uid)
			.get()
			.then(playerList => {
				if (playerList.size > 1) throw new Error('More than one user found');
				else if (playerList.size === 1) {
					resolve(playerList.docs[0].ref);
				} else {
					throw new Error('No users found');
				}
			})
			.catch(err => {
				console.error(err);
				reject(new Error('Getting current user DocumentReference'));
			});
	});
}

export function getGameDocRef(): firebase.firestore.DocumentReference {
	const db = firebase.firestore();
	const gameId = store.getState().main.gameId;
	if (!gameId)
		throw new Error('Could not get gameDocRef because gameId is not set');

	return db.doc(`games/${gameId}`);
}
