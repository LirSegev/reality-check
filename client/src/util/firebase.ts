import * as firebase from 'firebase/app';
import { getCurrentPlayer } from './db';

export function signOut(isAdmin?: boolean) {
	if (!isAdmin) {
		const gameId = localStorage.getItem('gameId');
		if (gameId)
			getCurrentPlayer().then(player => {
				if (player && player.messagingToken) {
					const removeDeviceFromDeviceGroup = firebase
						.functions()
						.httpsCallable('removeDeviceFromDeviceGroup');

					removeDeviceFromDeviceGroup({
						token: player.messagingToken,
						gameId,
					});
				} else {
					console.error(new Error("Player doesn't have messagingToken"));
				}
			});
		else console.error(new Error('Error no gameID'));
	}
	firebase.auth().signOut();
}
