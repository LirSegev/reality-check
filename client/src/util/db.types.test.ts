import * as firebase from 'firebase';
import { fold } from 'fp-ts/lib/Either';
import { pipe } from 'fp-ts/lib/pipeable';
import * as t from 'io-ts';
import { PathReporter } from 'io-ts/lib/PathReporter';

import { PlayerCodec } from './db.types';

function expectSuccess<T>(result: t.Validation<T>, expected?: T): void {
	pipe(
		result,
		fold(
			() => {
				throw new Error(`${result} is not a right`);
			},
			a => {
				if (expected !== undefined) {
					expect(a).toEqual(expected);
				}
			}
		)
	);
}

function expectFailure(
	result: t.Validation<any>,
	errors: Array<string>,
	matchSnapshot?: boolean | string
): void {
	pipe(
		result,
		fold(
			() => {
				const report = PathReporter.report(result);
				expect(report).toIncludeAllMembers(errors);
				if (matchSnapshot)
					expect(report).toMatchSnapshot(
						typeof matchSnapshot === 'string' ? matchSnapshot : undefined
					);
			},
			() => {
				throw new Error(`${result} is not a left`);
			}
		)
	);
}

describe('PlayerCodec', () => {
	describe('should succeed validating a valid value', () => {
		const T = PlayerCodec;

		test('without partials', () => {
			expectSuccess(
				T.decode({
					displayName: 'lir',
					role: 'chaser',
					uid: 'someUid',
					location: {
						geopoint: new firebase.firestore.GeoPoint(1, 2),
						timestamp: new firebase.firestore.Timestamp(1, 2),
					},
				})
			);
			expectSuccess(
				T.decode({
					displayName: 'test',
					role: 'intelligence',
					uid: 'anotherUid',
					location: null,
				})
			);
		});
		test('with partials', () => {
			expectSuccess(
				T.decode({
					displayName: 'lir',
					role: 'detective',
					uid: 'someUid',
					location: {
						geopoint: new firebase.firestore.GeoPoint(1, 2),
						timestamp: new firebase.firestore.Timestamp(1, 2),
					},
					isDeleted: true,
					messagingToken: 'messaging',
				})
			);
		});
	});

	it('should fail validating an invalid value', () => {
		const T = PlayerCodec;

		expectFailure(
			T.decode({
				messagingToken: 12345,
				isDeleted: 'yes',
			}),
			[
				'Invalid value undefined supplied to : Player/0: required/displayName: string',
				'Invalid value undefined supplied to : Player/0: required/uid: string',
				'Invalid value 12345 supplied to : Player/1: optional/messagingToken: string',
				'Invalid value "yes" supplied to : Player/1: optional/isDeleted: boolean',
			],
			true
		);
		expectFailure(
			T.decode({
				displayName: 123,
				role: 'lir',
				uid: null,
				location: { lat: 12, long: 12 },
			}),
			[
				'Invalid value 123 supplied to : Player/0: required/displayName: string',
				'Invalid value null supplied to : Player/0: required/uid: string',
				'Invalid value {"lat":12,"long":12} supplied to : Player/0: required/location: (PlayerLocation | null)/1: null',
			],
			true
		);
	});

	it('should not pass extra params', () => {
		const T = PlayerCodec;
		expect.assertions(2);

		expectSuccess(
			T.decode({
				displayName: 'lir',
				role: 'chaser',
				uid: 'someUid',
				location: {
					geopoint: new firebase.firestore.GeoPoint(1, 2),
					timestamp: new firebase.firestore.Timestamp(1, 2),
				},
				test: 'lir',
			}),
			{
				displayName: 'lir',
				role: 'chaser',
				uid: 'someUid',
				location: {
					geopoint: new firebase.firestore.GeoPoint(1, 2),
					timestamp: new firebase.firestore.Timestamp(1, 2),
				},
			}
		);
		expectSuccess(
			T.decode({
				displayName: 'lir',
				role: 'chaser',
				uid: 'someUid',
				location: {
					geopoint: new firebase.firestore.GeoPoint(1, 2),
					timestamp: new firebase.firestore.Timestamp(1, 2),
					test: 'lir',
				},
			}),
			{
				displayName: 'lir',
				role: 'chaser',
				uid: 'someUid',
				location: {
					geopoint: new firebase.firestore.GeoPoint(1, 2),
					timestamp: new firebase.firestore.Timestamp(1, 2),
				},
			}
		);
	});
});
