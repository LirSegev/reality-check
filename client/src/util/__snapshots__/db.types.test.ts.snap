// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerCodec should fail validating an invalid value 1`] = `
Array [
  "Invalid value undefined supplied to : Player/0: required/displayName: string",
  "Invalid value undefined supplied to : Player/0: required/role: PlayerRole",
  "Invalid value undefined supplied to : Player/0: required/uid: string",
  "Invalid value undefined supplied to : Player/0: required/location: (PlayerLocation | null)/0: PlayerLocation",
  "Invalid value undefined supplied to : Player/0: required/location: (PlayerLocation | null)/1: null",
  "Invalid value 12345 supplied to : Player/1: optional/messagingToken: string",
  "Invalid value \\"yes\\" supplied to : Player/1: optional/isDeleted: boolean",
]
`;

exports[`PlayerCodec should fail validating an invalid value 2`] = `
Array [
  "Invalid value 123 supplied to : Player/0: required/displayName: string",
  "Invalid value \\"lir\\" supplied to : Player/0: required/role: PlayerRole",
  "Invalid value null supplied to : Player/0: required/uid: string",
  "Invalid value undefined supplied to : Player/0: required/location: (PlayerLocation | null)/0: PlayerLocation/geopoint: FirebaseGeoPoint",
  "Invalid value undefined supplied to : Player/0: required/location: (PlayerLocation | null)/0: PlayerLocation/timestamp: FirebaseTimestamp",
  "Invalid value {\\"lat\\":12,\\"long\\":12} supplied to : Player/0: required/location: (PlayerLocation | null)/1: null",
]
`;
